src/Roslyn.Diagnostics.Analyzers/UnitTests/SymbolIsBannedAnalyzerTests.cs

```diff
@@ -0,0 +1,375 @@
+        {
+            return new SymbolIsBannedAnalyzer();
+        }
+
+        protected override DiagnosticAnalyzer GetBasicDiagnosticAnalyzer()
+        {
+            return new SymbolIsBannedAnalyzer();
+        }
+
+        private void VerifyBasic(string source, string bannedApiText, params DiagnosticResult[] expected)
+        {
+            var additionalFiles = GetAdditionalTextFiles(SymbolIsBannedAnalyzer.BannedSymbolsFileName, bannedApiText);
+            Verify(source, LanguageNames.VisualBasic, GetBasicDiagnosticAnalyzer(), additionalFiles, compilationOptions: null, parseOptions: null, expected: expected);
+        }
+
+        private void VerifyCSharp(string source, string bannedApiText, params DiagnosticResult[] expected)
+        {
+            var additionalFiles = GetAdditionalTextFiles(SymbolIsBannedAnalyzer.BannedSymbolsFileName, bannedApiText);
+            Verify(source, LanguageNames.CSharp, GetCSharpDiagnosticAnalyzer(), additionalFiles, compilationOptions: null, parseOptions: null, expected: expected);
+        }
+
+        #region Diagnostic tests
+
+        [Fact]
+        public void NoDiagnosticForNoBannedText()
+        {
+            Verify("class C { }", LanguageNames.CSharp, GetCSharpDiagnosticAnalyzer(), additionalFiles: null, compilationOptions: null, parseOptions: null, expected: Array.Empty<DiagnosticResult>());
+        }
+
+        [Fact]
+        public void NoDiagnosticReportedForEmptyBannedText()
+        {
+            var source = @"";
+
+            var bannedText = @"";
+
+            VerifyCSharp(source, bannedText);
+        }
+
+        [Fact]
+        public void DiagnosticReportedForDuplicateBannedApiLines()
+        {
+            var source = @"";
+            var bannedText = @"
+System.Console
+System.Console";
+
+            VerifyCSharp(source, bannedText,
+                GetResultAt(
+                    SymbolIsBannedAnalyzer.BannedSymbolsFileName,
+                    SymbolIsBannedAnalyzer.DuplicateBannedSymbolRule.Id,
+                    string.Format(SymbolIsBannedAnalyzer.DuplicateBannedSymbolRule.MessageFormat.ToString(), "System.Console"),
+                    "(3,1)", "(2,1)"));
+        }
+
+        [Fact]
+        public void CSharp_DiagnosticReportedForTypeInSource()
+        {
+            var source = @"
+namespace N
+{
+    class Banned { }
+    class C
+    {
+        void M()
+        {
+            var c = new Banned();
+        }
+    }
+}";
+
+            var bannedText = @"
+N.Banned";
+
+            VerifyCSharp(source, bannedText, GetCSharpResultAt(9, 21, SymbolIsBannedAnalyzer.SymbolIsBannedRule, "Banned"));
+        }
+
+        [Fact]
+        public void CSharp_DiagnosticReportedForGenericTypeFromMetadata()
+        {
+            var source = @"
+class C
+{
+    void M()
+    {
+        var c = new System.Collections.Generic.List<string>();
+    }
+}";
+
+            var bannedText = @"
+System.Collections.Generic.List`1";
```

Comment:
@Pilchie Are the developers expected to generate the API file entry manually, i.e. without a code fix? Is it worth considering offering some kind of code fix, for e.g. if the API file has a temporary line with just the symbol name, then a hidden diagnostic is reported on all symbol declarations with matching name and a code fix replaces the symbol name entry in the API file with the valid entry. Not sure if there are any better approaches.

Reply:
I was assuming people would just type it, but agree that it might be worth considering something else.

Reply:
Especially if we plan to extend it in future to allow banned method/properties with parameters.

Reply:
Is there a reason we could not add another diagnostic for the banned API text file, with codefix?

Reply:
@drewnoakes the question is what do you generate the diagnostic on?  Everything that isn't currently banned?

Reply:
I meant that the new diagnostic would only apply in "BannedAPIs.txt". It would provide assistance when the ban list is being edited by flagging doc comment IDs that can't be resolved to known symbols.

As for a code fix, even just opening a link that explains the doc comment ID syntax would be useful.

It'd be annoying to have a "ban this API" entry on every screwdriver menu. How else would we surface a "ban this" command in the IDE?

Reply:
> Is there a reason we could not add another diagnostic for the banned API text file, with codefix?

@drewnoakes That would be ideal, but unfortunately we do not support code fixes/light bulb in additional files, so the diagnostic has to be in regular source code. My suggestion was that whenever user has an incomplete entry in banned API file, say an entry with just the symbol name instead of entire documentation comment ID `Member1`, every symbol in the compilation with  name `Member1` will have a hidden diagnostic that offers to add an entry for it in the banned API file and the user can go back to source and invoke the code fix on the `Member1` that they want to ban.

Regardless, we can always wait for more dogfooding feedback before implementing anything here.

Reply:
Actually, my suggestion will not work if you want to ban APIs from referenced projects/assemblies, so we probably need a different approach.